File: app/src/core/runbook/cron.ts
/**
 * Runbooks — a small 5-field cron parser for the Schedules UI (R4b). It mirrors
 * `backend/internal/orchestrator/cron.go`: the backend is authoritative for
 * firing; this is used to validate the expression as the user types and to
 * preview the next few run times. Framework-free.
 *
 * Fields: minute hour day-of-month month day-of-week. Each accepts a wildcard,
 * a single value, ranges, step values, and comma lists. Day-of-week is 0-6
 * (Sun = 0; 7 also accepted). Macros: @hourly @daily @weekly @monthly @yearly.
 * Times are computed in the runtime's local zone, matching the Go backend.
 */

const MACROS: Record<string, string> = {
  '@yearly': '0 0 1 1 *',
  '@annually': '0 0 1 1 *',
  '@monthly': '0 0 1 * *',
  '@weekly': '0 0 * * 0',
  '@daily': '0 0 * * *',
  '@midnight': '0 0 * * *',
  '@hourly': '0 * * * *',
}

interface ParsedCron {
  minute: Set<number>
  hour: Set<number>
  dom: Set<number>
  month: Set<number>
  dow: Set<number>
  domRestricted: boolean
  dowRestricted: boolean
}

function parseField(field: string, min: number, max: number): Set<number> {
  const out = new Set<number>()
  for (const rawPart of field.split(',')) {
    const part = rawPart.trim()
    if (!part) throw new Error('empty term')
    let step = 1
    let range = part
    const slash = part.indexOf('/')
    if (slash >= 0) {
      range = part.slice(0, slash)
      step = Number(part.slice(slash + 1))
      if (!Number.isInteger(step) || step <= 0) throw new Error(`bad step "${part.slice(slash + 1)}"`)
    }
    let lo = min
    let hi = max
    if (range !== '*') {
      const dash = range.indexOf('-')
      if (dash >= 0) {
        lo = Number(range.slice(0, dash))
        hi = Number(range.slice(dash + 1))
      } else {
        lo = hi = Number(range)
      }
      if (!Number.isInteger(lo) || !Number.isInteger(hi)) throw new Error(`bad value "${range}"`)
    }
    if (lo < min || hi > max || lo > hi) throw new Error(`out of range [${min}-${max}]: "${part}"`)
    for (let v = lo; v <= hi; v += step) out.add(v)
  }
  return out
}

/** Parse a cron expression. Throws `Error` with a human message on bad input. */
export function parseCron(expr: string): ParsedCron {
  const raw = expr.trim()
  if (!raw) throw new Error('empty cron expression')
  const spec = MACROS[raw.toLowerCase()] ?? raw
  const f = spec.split(/\s+/)
  if (f.length !== 5) throw new Error(`cron needs 5 fields (got ${f.length})`)
  const dowNorm = f[4]
    .split(',')
    .map((p) => p.split('-').map((x) => (x.trim() === '7' ? '0' : x)).join('-'))
    .join(',')
  const wrap = (label: string, fn: () => Set<number>) => {
    try {
      return fn()
    } catch (e) {
      throw new Error(`${label}: ${e instanceof Error ? e.message : e}`)
    }
  }
  return {
    minute: wrap('minute', () => parseField(f[0], 0, 59)),
    hour: wrap('hour', () => parseField(f[1], 0, 23)),
    dom: wrap('day-of-month', () => parseField(f[2], 1, 31)),
    month: wrap('month', () => parseField(f[3], 1, 12)),
    dow: wrap('day-of-week', () => parseField(dowNorm, 0, 6)),
    domRestricted: f[2] !== '*',
    dowRestricted: f[4] !== '*',
  }
}

/** True if `expr` is a valid cron expression. */
export function isValidCron(expr: string): boolean {
  try {
    parseCron(expr)
    return true
  } catch {
    return false
  }
}

function dayMatches(p: ParsedCron, d: Date): boolean {
  const dom = p.dom.has(d.getDate())
  const dow = p.dow.has(d.getDay())
  if (p.domRestricted && p.dowRestricted) return dom || dow
  if (p.domRestricted) return dom
  if (p.dowRestricted) return dow
  return true
}

/** The first time strictly after `after` that matches, or null if none within ~4y. */
export function cronNext(expr: string, after: Date = new Date()): Date | null {
  const p = parseCron(expr)
  const t = new Date(after.getTime())
  t.setSeconds(0, 0)
  t.setMinutes(t.getMinutes() + 1)
  const limit = new Date(after.getTime())
  limit.setFullYear(limit.getFullYear() + 4)
  while (t < limit) {
    if (!p.month.has(t.getMonth() + 1)) {
      t.setMonth(t.getMonth() + 1, 1)
      t.setHours(0, 0, 0, 0)
      continue
    }
    if (!dayMatches(p, t)) {
      t.setDate(t.getDate() + 1)
      t.setHours(0, 0, 0, 0)
      continue
    }
    if (!p.hour.has(t.getHours())) {
      t.setHours(t.getHours() + 1, 0, 0, 0)
      continue
    }
    if (!p.minute.has(t.getMinutes())) {
      t.setMinutes(t.getMinutes() + 1, 0, 0)
      continue
    }
    return t
  }
  return null
}

/** The next `count` run times, for a preview. */
export function cronNextRuns(expr: string, count = 3, after: Date = new Date()): Date[] {
  const out: Date[] = []
  let cur = after
  for (let i = 0; i < count; i++) {
    const n = cronNext(expr, cur)
    if (!n) break
    out.push(n)
    cur = n
  }
  return out
}

const COMMON: Record<string, string> = {
  '@hourly': 'every hour',
  '@daily': 'every day at midnight',
  '@weekly': 'every Sunday at midnight',
  '@monthly': 'the 1st of every month',
  '@yearly': 'every January 1st',
  '* * * * *': 'every minute',
  '0 * * * *': 'every hour, on the hour',
  '0 0 * * *': 'every day at midnight',
  '0 9 * * 1-5': 'weekdays at 09:00',
  '*/5 * * * *': 'every 5 minutes',
  '*/15 * * * *': 'every 15 minutes',
  '*/30 * * * *': 'every 30 minutes',
}

/** A short human description — falls back to the raw expression. */
export function describeCron(expr: string): string {
  const raw = expr.trim().toLowerCase()
  if (COMMON[raw]) return COMMON[raw]
  const m = /^(\d{1,2}) (\d{1,2}) \* \* \*$/.exec(raw)
  if (m) return `every day at ${m[2].padStart(2, '0')}:${m[1].padStart(2, '0')}`
  return expr.trim()
}

File: backend/internal/orchestrator/cron.go
package orchestrator

import (
	"fmt"
	"strconv"
	"strings"
	"time"
)

// CronExpr is a parsed 5-field cron expression: minute hour day-of-month month
// day-of-week. Fields accept `*`, `a`, `a-b`, `a-b/n`, `*/n`, and comma lists of
// those. Day-of-week is 0-6 (Sunday = 0; 7 also accepted for Sunday). The
// convenience macros `@hourly`, `@daily`, `@weekly`, `@monthly`, `@yearly` are
// also understood. This is a small self-contained parser — no external cron
// dependency (RUNBOOK_MODULE_PLAN.md §6.4: no new Go deps).
type CronExpr struct {
	raw    string
	minute [60]bool
	hour   [24]bool
	dom    [32]bool // 1-31
	month  [13]bool // 1-12
	dow    [7]bool  // 0-6
	// domRestricted / dowRestricted follow Vixie cron semantics: when only one
	// of the two day fields is a wildcard, the other alone decides the day.
	domRestricted bool
	dowRestricted bool
}

var cronMacros = map[string]string{
	"@yearly":   "0 0 1 1 *",
	"@annually": "0 0 1 1 *",
	"@monthly":  "0 0 1 * *",
	"@weekly":   "0 0 * * 0",
	"@daily":    "0 0 * * *",
	"@midnight": "0 0 * * *",
	"@hourly":   "0 * * * *",
}

// ParseCron parses a 5-field expression (or a macro).
func ParseCron(expr string) (CronExpr, error) {
	raw := strings.TrimSpace(expr)
	if raw == "" {
		return CronExpr{}, fmt.Errorf("empty cron expression")
	}
	spec := raw
	if m, ok := cronMacros[strings.ToLower(raw)]; ok {
		spec = m
	}
	fields := strings.Fields(spec)
	if len(fields) != 5 {
		return CronExpr{}, fmt.Errorf("cron expression must have 5 fields (got %d)", len(fields))
	}
	c := CronExpr{raw: raw}
	if err := fillField(fields[0], 0, 59, c.minute[:]); err != nil {
		return CronExpr{}, fmt.Errorf("minute: %w", err)
	}
	if err := fillField(fields[1], 0, 23, c.hour[:]); err != nil {
		return CronExpr{}, fmt.Errorf("hour: %w", err)
	}
	if err := fillField(fields[2], 1, 31, c.dom[:]); err != nil {
		return CronExpr{}, fmt.Errorf("day-of-month: %w", err)
	}
	if err := fillField(fields[3], 1, 12, c.month[:]); err != nil {
		return CronExpr{}, fmt.Errorf("month: %w", err)
	}
	if err := fillField(normalizeDow(fields[4]), 0, 6, c.dow[:]); err != nil {
		return CronExpr{}, fmt.Errorf("day-of-week: %w", err)
	}
	c.domRestricted = fields[2] != "*"
	c.dowRestricted = fields[4] != "*"
	return c, nil
}

// String returns the original (pre-macro) expression.
func (c CronExpr) String() string { return c.raw }

// normalizeDow maps the Vixie "7 = Sunday" convention onto 0 for whole tokens
// (a bare "7" or the endpoints of a range), without mangling multi-digit text.
func normalizeDow(field string) string {
	repl := func(tok string) string {
		if tok == "7" {
			return "0"
		}
		return tok
	}
	parts := strings.Split(field, ",")
	for i, p := range parts {
		if slash := strings.IndexByte(p, '/'); slash >= 0 {
			parts[i] = normalizeDowRange(p[:slash], repl) + "/" + p[slash+1:]
		} else {
			parts[i] = normalizeDowRange(p, repl)
		}
	}
	return strings.Join(parts, ",")
}

func normalizeDowRange(rng string, repl func(string) string) string {
	if dash := strings.IndexByte(rng, '-'); dash >= 0 {
		return repl(rng[:dash]) + "-" + repl(rng[dash+1:])
	}
	return repl(rng)
}

func fillField(field string, min, max int, set []bool) error {
	for _, part := range strings.Split(field, ",") {
		part = strings.TrimSpace(part)
		if part == "" {
			return fmt.Errorf("empty term")
		}
		step := 1
		rng := part
		if slash := strings.IndexByte(part, '/'); slash >= 0 {
			rng = part[:slash]
			s, err := strconv.Atoi(part[slash+1:])
			if err != nil || s <= 0 {
				return fmt.Errorf("bad step %q", part[slash+1:])
			}
			step = s
		}
		lo, hi := min, max
		if rng != "*" {
			if dash := strings.IndexByte(rng, '-'); dash >= 0 {
				a, err1 := strconv.Atoi(rng[:dash])
				b, err2 := strconv.Atoi(rng[dash+1:])
				if err1 != nil || err2 != nil {
					return fmt.Errorf("bad range %q", rng)
				}
				lo, hi = a, b
			} else {
				n, err := strconv.Atoi(rng)
				if err != nil {
					return fmt.Errorf("bad value %q", rng)
				}
				lo, hi = n, n
			}
		}
		if lo < min || hi > max || lo > hi {
			return fmt.Errorf("value out of range [%d-%d]: %q", min, max, part)
		}
		// set is value-indexed (set[v]); the arrays are sized to allow it.
		for v := lo; v <= hi; v += step {
			set[v] = true
		}
	}
	return nil
}

func anyTrue(s []bool) bool {
	for _, v := range s {
		if v {
			return true
		}
	}
	return false
}

// Next returns the first time strictly after `after` that matches the
// expression, truncated to the minute. It gives up (returns zero) after
// scanning ~4 years, which only happens for an impossible date like Feb 30.
func (c CronExpr) Next(after time.Time) time.Time {
	if !anyTrue(c.minute[:]) { // zero value / never parsed
		return time.Time{}
	}
	t := after.Truncate(time.Minute).Add(time.Minute)
	limit := after.AddDate(4, 0, 0)
	for t.Before(limit) {
		if !c.month[t.Month()] {
			// jump to the first day of the next month
			t = time.Date(t.Year(), t.Month(), 1, 0, 0, 0, 0, t.Location()).AddDate(0, 1, 0)
			continue
		}
		if !c.dayMatches(t) {
			t = time.Date(t.Year(), t.Month(), t.Day(), 0, 0, 0, 0, t.Location()).AddDate(0, 0, 1)
			continue
		}
		if !c.hour[t.Hour()] {
			t = time.Date(t.Year(), t.Month(), t.Day(), t.Hour(), 0, 0, 0, t.Location()).Add(time.Hour)
			continue
		}
		if !c.minute[t.Minute()] {
			t = t.Add(time.Minute)
			continue
		}
		return t
	}
	return time.Time{}
}

func (c CronExpr) dayMatches(t time.Time) bool {
	dom := c.dom[t.Day()]
	dow := c.dow[int(t.Weekday())]
	switch {
	case c.domRestricted && c.dowRestricted:
		return dom || dow // Vixie cron: union when both are restricted
	case c.domRestricted:
		return dom
	case c.dowRestricted:
		return dow
	default:
		return true
	}
}
